typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1005 @@
+
+### Peano numbers
+
+Moving deeper, we can think about type-level programming and, more concretely, about computational on types.
+The basic primitive which may help us in this problem is Peano numbers.
+Lets see how it may look in TypeScript.
+
+```typescript
+type Zero = "zero";
+
+type Nat = Zero | { n: Nat };
+```
+
+And now we are able to define `Succ`, which adds one to our `Nat`.
+
+```typescript
+type Succ<N extends Nat> = { n: N };
+```
+
+With a simple `Nat` definition, we are able to make something useful.
+
+Let's try to create a type-safe vector that will store its length in type.
+
+Empty vector will be simply be a `"nil"`.
+
+```typescript
+type Nil = "nil";
+```
+Type `Cons` will add a value to the vector and increase its type-level length.
+
+```typescript
+type Cons<A, N extends Nat> = {a: A, v: Vec<A, N>};
+```
+
+And the main type `Vec` combines them with conditional type.
+
+Using `infer`, we are able to infer new type variable `R` from `N`.
+So, when `N` is a `Succ`, we can infer nested `R` from it and return `Cons<A, R>`.
+Otherwise, it is simply `Nil`.
+
+```typescript
+type Vec<A, N extends Nat> = N extends Succ<infer R> ? Cons<A, R> : Nil;
+```
+
+And here are examples of `Vec`'s usage.
+We define two helper function which build `Vec` values.
+And, as you can see, we can use it safely.
+
+```typescript
+const emptyVec: <A>() => Vec<A, Zero> = () => "nil";
+
+const pushVec: <A, N extends Nat>(a: A, v: Vec<A, N>) => Vec<A, Succ<N>> = (a, v) => {
+  return {a, v}
+};
+
+let empty: Vec<number, Zero> = emptyVec(); // Ok
+let oneElem: Vec<number, Succ<Zero>> = pushVec(1, empty); // Ok
+let twoElems: Vec<number, Succ<Succ<Zero>>> = pushVec(2, oneElem); // Ok
+let twoElemsInvalid: Vec<number, Succ<Zero>> = pushVec(2, oneElem); // error
+oneElem = twoElems; // error
+```
+
+This was a simple example what we can do with type level programming.
+Using such primitives, we can implement different operations on types and even a Fibonacci [implementation](https://mjj.io/2021/03/29/type-level-programming-in-typescript/), for example.
+Follow Mathias Jean Johansen's [blog post](https://mjj.io/2021/03/29/type-level-programming-in-typescript/) or above-mentioned Yuriy Bogomolov's [examples](https://github.com/YBogomolov/talk-typelevel-ts/blob/master/src/typelevel.d.ts).
+
+```typescript
+type Fibonacci<N, F0 = Zero, F1 = One> = {
+  acc: F0
+  n: N extends Succ<infer _> ? Fibonacci<Decrement<N>, F1, Add<F0, F1>> : never
+}[IfElse<Equals<Zero, N>, "acc", "n">]
+```
+
+### GADT and eDSL
+
+#### GADT
+
+Generalized algebraic data types in Haskell give us an ability to manually write types of the constructors.
+Having data type `D a` with type value `a` we are able to create constructors like `C :: Int -> D Int`.
+The key feature here is that Haskell may say us about equality of types `a` and `Int` here.
```

Comment:
This line is a bit oddly worded, could you try to rephrase. In particular, "Haskell may say us".